### PR TITLE
Support camelCase shortcuts for configuration selection in dependencies/dependencyInsight task

### LINKED ISSUE #16141: Support camelCase shourtcuts for configuration selection in dependencies/dependencyInsinght task
Currently `--configuration` parameter in `dependencies` and `dependencyInsight` tasks requires user to type full configuration name.

I propose add `NameMatcher`-based support for camelCase in command line parameter to improve CLI UX like it done for projects and tasks selection.

In case of ambiguous configuration selected by camel case task should fail. 

### Expected Behavior
`gradle dependencies --configuration ruC` should work same as `gradle dependencies --configuration runtimeClasspath` if `ruC` there's only one configuration with such abbrev.

### Current Behavior
Configuration not found

### Context
Current zsh gradle autocompletion don't support configurations (and it would be be quite costly to do so) and when you investigate dependency resolution problems it would be useful.

Also I brought idea in gradle community [slack #general][1] and it found some support. 

[1]: https://gradle-community.slack.com/archives/CA7UM03V3/p1611706682059700

### PR DESCRIPTION
**Pull request by grossws**
Fixes #16141

I'll add relevant docs for this feature if required.

Do I need to add some integration tests for such trivial change? I've added couple of tests into `DependencyReportTaskTest` which check relevant behavior change.

Also `sanityCheck` against master fails 'cause dependency verification failures in independent modules. So I checked relevant todo item below.

### Context
See #16141 and https://gradle-community.slack.com/archives/CA7UM03V3/p1611706682059700

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [x] Recognize contributor in release notes

### CODE REVIEW

**Comment by grossws** [comment]
Rebased on master and squashed commits into one

**Comment by donat** [comment]
@grossws Sorry for the late reply. Your PR makes sense. I'll give it a more thorough review.

File: subprojects/diagnostics/src/test/groovy/org/gradle/api/tasks/diagnostics/DependencyReportTaskTest.groovy [hunk 1]
```diff
@@ -85,4 +86,28 @@ class DependencyReportTaskTest extends AbstractProjectBuilderSpec {
         then:
         task.configurations == [bConf] as Set
     }
+
+    def "configuration to render could be specified by camelCase shortcut"() {
```
Comment (donat) [inline subprojects/diagnostics/src/test/groovy/org/gradle/api/tasks/diagnostics/DependencyReportTaskTest.groovy:90]: ❌ Please add a similar coverage to the `dependencyInsight` task. Maybe add it to the `DependencyInsightReportTaskIntegrationTest` class.

File: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/viewing_debugging_dependencies.adoc [hunk 2]
```diff
@@ -49,6 +49,8 @@ For example, to show dependencies that would be on the test runtime classpath in
 gradle -q dependencies --configuration testRuntimeClasspath
 ----
 
+TIP: You can use _camelCase_ shortcut to select configuration like `tRC` instead of `testRuntimeClasspath` if that shortcut points to unique configuration.
```
Comment (donat) [inline subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/viewing_debugging_dependencies.adoc:52]: Please mention this change in the `subprojects/docs/src/docs/release/notes.md`.

**Review by donat — CHANGES_REQUESTED** [review]
Thanks for the quality pull request. I have just a couple of minor requests, otherwise, it's good for merging.

**Comment by donat** [comment]
Running the changes through CI: https://builds.gradle.org/viewQueued.html?itemId=43971909

File: subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/ConfigurationFinder.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.diagnostics.internal;
+
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.ConfigurationContainer;
+import org.gradle.util.NameMatcher;
```
Comment (donat) [inline subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/ConfigurationFinder.java:22]: ❌ Please use `org.gradle.util.internal.NameMatcher`.

**Comment by grossws** [comment]
@donat, thanks for review. I'll update PR when have a bit of time.

**Comment by donat** [comment]
Sounds good. Looking forward to the update.

**Comment by grossws** [comment]
@donat, I updated PR but I'm not quite sure what format should be used in `notes.md` for feature description.

**Comment by donat** [comment]
@grossws It's fine, I've adjusted the text there.

**Comment by donat** [comment]
@bot-gradle test and merge

**Comment by grossws** [comment]
Thanks, @donat

**Comment by lacasseio** [comment]
This PR breaks configurations that are created via Rules. There should be a final `getByName` if the finder here fails (https://github.com/gradle/gradle/commit/c190a1ba9997db532884c0f0a443dc4b29c638d2#diff-9c98c9ba8979d3a753ca3ebdc39a2c5d84bef93c3889132cc0ff67bf8702eedaR30). For example, in Nokee we use [this code](https://github.com/nokeedev/gradle-native/blob/ef45487ffe820073c082da016c79bec3bad74e41/subprojects/platform-jni/src/main/java/dev/nokee/platform/jni/internal/plugins/JniLibraryPlugin.java#L319-L321) to execute additional code that may create configurations. The [previous `getByName` here](https://github.com/gradle/gradle/commit/c190a1ba9997db532884c0f0a443dc4b29c638d2#diff-3abfca496347f99852cc4d194aa44ced6f255ab250be6813eabd667471f61d68L153) would trigger any container rules.

**Comment by donat** [comment]
@grossws https://github.com/gradle/gradle/pull/16142#issuecomment-859076187 is valid. I'm about to revert your PR: https://github.com/gradle/gradle/pull/17428.
Could you please submit another PR that takes the rules into account? 

An integtest is pretty simple for this case: `gradle dependencies --configuration foo` should pass for the following build script:

```groovy
project.afterEvaluate { p ->
    p.configurations.addRule('my custom config defined via a rule', { s -> 
        if (s == 'foo') {
            p.configurations.create('foo')
       }
	)
}
```

**Comment by grossws** [comment]
@donat, ok, I'll fix that and add tests. Should I create PR against current master or this PR branch?

@lacasseio, thanks for your feedback and catching a bug)

**Comment by lacasseio** [comment]
I feel bad for forcing a revert as the feature is pretty cool. I hope you can submit a change quickly. :-)

**Comment by grossws** [comment]
@lacasseio, already done in #17445 ,)

**Pull request merged**